developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,255 @@
+.. _virtualisation:
+
+SOF VirtIO design
+#################
+
+.. contents::
+
+1. General information
+**********************
+
+VirtIO is a platform-agnostic virtualisation standard, managed by OASIS
+https://www.oasis-open.org/committees/tc_home.php?wg_abbrev=virtio . It defines
+a design and a protocol for implementing device IO in VMs.
+
+When Linux is running in a VM, multiple drivers are already available, using
+virtio: networking, storage, etc. Those drivers use the virtio and virtqueue
+APIs to communicate with the host or, potentially, with other guests.
+
+The host part depends on the used virtualisation solution. Some of them
+implement the VirtIO API themselves, then using their method of choice to talk
+to the whatever host they are running on. A typical Linux virtualisation
+implementation uses QEMU + KVM to run guest instances. In this case there are
+two ways to implement a VirtIO host: in the user space and in the kernel. If
+implemented in the user space, it is managed by QEMU. In the kernel space VirtIO
+host drivers use the Linux vhost API.
+
+It is also worth to note, that VirtIO guest drivers reside in respective driver
+subdirectories in the Linux kernel, e.g. drivers/net, drivers/block etc. On the
+contrary, all vhost drivers are located under drivers/vhost. That directory also
+contains vhost.c and vhost.h which are used by all vhost drivers.
+
+The SOF virtualisation solution also implements a vhost driver.
+
+2. SOF VirtIO design
+********************
+
+The principal point of the SOF VirtIO design is using the IPC API, that is also
```

Comment:
principal point is to virtualize the IPC command and audio data flows between host and guests. I would introduce the doorbell, mailbox and zero copy (for audio data and stream position) concepts here too

Reply:
which IPC? the SOF-defined IPC?

Reply:
@lgirdwood I'm not sure we need the mailbox-doorbell concept here. In my understanding the mailbox-doorbell interface is symmetric, at least that's (at least conceptually) what it is like with SOF. VirtQueues OTOH are asymmetric. In fact only one side can be the initiator. That's why we have to pre-queue position update buffers from the guest. The opposite direction is always a response.

Reply:
@plbossart yes, SOF-defined

Reply:
we do, we need to explain that we are virtualising a mailbox and doorbell. we are not virtualising PCMs or kcontrols as some people may expect.

Reply:
ok, got it